### PR TITLE
[dynamic shape] Add TensorSpec for per-dimension dynamic shape specification

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #180923

## Summary

Adds `TensorSpec` — a list-like container of `IntSpec | None` with length equal to the tensor's number of dims. `None` entries inherit the default dynamism policy; non-`None` entries carry a per-dim `IntSpec`.

```python
from torch._dynamo.dynamic_spec import IntSpec, TensorSpec

ts = TensorSpec(2).dim(0, IntSpec.backed("batch")).dim(1, IntSpec.static())
```

Data class only — no `torch.compile` / `torch.export` integration in this PR. 

### Construction — three forms

```python
TensorSpec(2)                                 # int → 2 dims, all None
TensorSpec([IntSpec.backed("batch"), None])   # list/tuple → length = dim, entries used as-is
TensorSpec({0: IntSpec.backed("batch")})      # dict → dim = max(keys) + 1, missing keys = None
```

The constructor only type-dispatches: anything that isn't `int` / `list` / `tuple` / `dict` raises `TypeError`. Internal value/key validity (negative dim, empty dict, non-`IntSpec` entries) is the caller's responsibility — no granular checks.

### Mutability and access

- **Fluent `.dim(index, spec)`** — sets the per-dim spec, returns `self` for chaining.
- **List-like:** `ts[i]`, `ts[i] = spec`, `len(ts)`, `iter(ts)`.
- **No public read property for the dim count.** Use `len(ts)`; tests touching internals read `ts._dim` directly.
- **`repr(ts)`** — list-form (`TensorSpec([IntSpec(...), None, IntSpec(...)])`); shows every dim in order, including `None` entries.

### Consistency with `IntSpec`

- **No `__eq__` / `__hash__`.** Same call as PR 1 — specs are compile-time inputs, value-based equality would force cache keys to drift with object identity and conflict with the AOT-snapshot invariant.
- **Narrow public read surface.** Matches PR 1: only writes are public (the `.dim(i, spec)` fluent setter and `__setitem__`); reads go through `__len__` / `__iter__` / `__getitem__` / `_dim`. No standalone `.dim` getter property — the same name is the setter.
- **Naming.** `dim` (not `rank`) for the dim count — `rank` collides with distributed-training terminology.

### Test scaffolding canonical form

The PR 1 `_compile_with_dynamic_shapes` helper is simplified in this PR: it now expects a `TensorSpec` per tensor argument (anything else is skipped). All `TestIntSpecCompile` call sites that previously used the bare-dict form `{"x": {0: IntSpec.backed("batch")}}` are updated to `{"x": TensorSpec({0: IntSpec.backed("batch")})}`. There is exactly one canonical container.

## Test plan — `python test/dynamo/test_dynamic_spec.py`

Adds two classes alongside PR 1's six:

- **`TestTensorSpecConstruction`** — int / list / tuple / dict construction (with the new polymorphic `__init__`); `__getitem__` / `__setitem__`; fluent `.dim(i, spec)`; iteration; index-out-of-range; sparse per-dim set; `repr` round-trip via list-form match; rejects unsupported input type.

- **`TestTensorSpecCompile`** — `@skipIfTorchDynamo()` at class level. Four graph-inspecting tests:
  - `test_tensorspec_backed_graph_has_backed_symbol` — int + `.dim()` form: BACKED dim yields a backed SymInt; single compile across varying sizes.
  - `test_tensorspec_list_init_recompile_progression` — list-form construction: BACKED dim 0 absorbs shape changes; STATIC dim 1 forces recompile per distinct value; final graph has backed SymInt at dim 0 and concrete int at dim 1.
  - `test_tensorspec_dict_init_recompile_progression` — same recompile pattern via dict-form construction.
  - `test_tensorspec_none_entry_inherits_automatic_dynamic` — `None` entry doesn't mark the dim; falls through to dynamo's auto-dynamic (`automatic_dynamic_shapes=True`, pinned via `@torch._dynamo.config.patch`): specialize on first call, promote to backed on first variation. Distinguishes `None` from STATIC.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

File: test/dynamo/test_dynamic_spec.py [hunk 1]
```diff
@@ -36,7 +36,9 @@ def _tensor_placeholder_shape(gm):
 
 
 def _apply_intspec_to_tensor(tensor, shape_spec):
-    if isinstance(shape_spec, dict):
+    if isinstance(shape_spec, TensorSpec):
+        items = enumerate(shape_spec)
+    elif isinstance(shape_spec, dict):
         items = shape_spec.items()
     elif isinstance(shape_spec, (list, tuple)):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: no need to support list and tuple anymore or dict now we have Tensor Spec we shall be strict with forcing using it. also remove tests that accept lists/dict for TensorSpec

File: torch/_dynamo/dynamic_spec.py [hunk 2]
```diff
@@ -302,3 +303,63 @@ def __repr__(self) -> str:
...
+    the compile context.
+
+    Example::
+
+        ts = TensorSpec(3)
+        ts.set(0, IntSpec.backed("batch", min=1, max=64))
+        # dims 1 and 2 are None -> inherit context default
+    """
+
+    def __init__(self, rank: int) -> None:
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py]: I suggest we support he following as input to constructs
arg can be 

1. size

2. list 
TensorSpec([IntSpec,…])

3. dict
TensorSpec({index->IntSpec)})

File: torch/_dynamo/dynamic_spec.py [hunk 3]
```diff
@@ -302,3 +303,63 @@ def __repr__(self) -> str:
...
+        """Construct from an existing list of specs."""
+        ts = cls(len(specs))
+        ts._specs = list(specs)
+        return ts
+
+    @property
+    def rank(self) -> int:
+        return self._rank
+
+    def set(self, index: int, spec: IntSpec) -> "TensorSpec":
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py]: use .dim(index, ..)

File: torch/_dynamo/dynamic_spec.py [hunk 4]
```diff
@@ -302,3 +303,63 @@ def __repr__(self) -> str:
...
+        """Set the spec at ``index`` and return ``self`` for chaining."""
+        self._specs[index] = spec
+        return self
+
+    def __getitem__(self, index: int) -> IntSpec | None:
+        return self._specs[index]
+
+    def __setitem__(self, index: int, spec: IntSpec | None) -> None:
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py:351]: are we supporting TensorSpec[0] = IntSpec here? thats fine. sounds good.

**Review by laithsakka — CHANGES_REQUESTED** [review]

**Comment by laithsakka** [comment]
looks good overall. left some comments to address

File: torch/_dynamo/dynamic_spec.py (hunk 4 above)
Comment (fxdawnn) [reply torch/_dynamo/dynamic_spec.py:351]: yes!

File: torch/_dynamo/dynamic_spec.py (hunk 3 above)
Comment (fxdawnn) [reply torch/_dynamo/dynamic_spec.py]: updated!

File: torch/_dynamo/dynamic_spec.py [hunk 5]
```diff
@@ -292,3 +293,75 @@ def __repr__(self) -> str:
...
+
+    Example::
+        TensorSpec(3)  # rank 3, all None
+        TensorSpec([IntSpec.backed("batch"), None])  # rank 2, dim 0 backed
+        TensorSpec({0: IntSpec.backed("batch")})  # rank 1, dim 0 backed
+    """
+
+    def __init__(
+        self,
+        spec: int
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py]: nit spec -> arg

File: test/dynamo/test_dynamic_spec.py [hunk 6]
```diff
@@ -418,6 +411,189 @@ def test_field_type_rejected(self, field, bad, factory_name, type_name, entry):
...
+
+@skipIfTorchDynamo()
+class TestTensorSpecCompile(TestCase):
+    """TensorSpec + compile integration via the test-local
+    `_compile_with_dynamic_shapes` helper. Same scaffolding path as
+    `TestIntSpecCompile` — the helper walks per-dim ``IntSpec`` from
+    the TensorSpec and applies ``mark_*`` to the tensor.
+    """
+
+    def test_tensorspec_backed_graph_has_backed_symbol(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: not needed i think you already have a test for this In previous PR

File: test/dynamo/test_dynamic_spec.py [hunk 7]
```diff
@@ -418,6 +411,189 @@ def test_field_type_rejected(self, field, bad, factory_name, type_name, entry):
...
+            backend=backend,
+        )
+        for n in [4, 8, 16, 32]:
+            fn(torch.randn(n, 3))
+        self.assertEqual(len(backend.graphs), 1)
+        shape = _tensor_placeholder_shape(backend.graphs[0])
+        self.assertIsInstance(shape[0], torch.SymInt)
+        self.assertEqual(len(free_unbacked_symbols(shape[0])), 0)
+
+    def test_tensorspec_list_init_recompile_progression(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:487]: already tested befoee

File: test/dynamo/test_dynamic_spec.py [hunk 8]
```diff
@@ -418,6 +411,189 @@ def test_field_type_rejected(self, field, bad, factory_name, type_name, entry):
...
+        fn(torch.randn(16, 7))
+        self.assertEqual(len(backend.graphs), 3)
+
+        shape = _tensor_placeholder_shape(backend.graphs[-1])
+        self.assertIsInstance(shape[0], torch.SymInt)
+        self.assertEqual(len(free_unbacked_symbols(shape[0])), 0)
+        self.assertIsInstance(shape[1], int)
+        self.assertEqual(shape[1], 7)
+
+    def test_tensorspec_dict_init_recompile_progression(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py]: not needed

File: test/dynamo/test_dynamic_spec.py [hunk 9]
```diff
@@ -418,6 +411,189 @@ def test_field_type_rejected(self, field, bad, factory_name, type_name, entry):
...
+        fn(torch.randn(8, 3))
+        fn(torch.randn(16, 3))
+        self.assertEqual(len(backend.graphs), 1)
+        fn(torch.randn(16, 5))
+        self.assertEqual(len(backend.graphs), 2)
+        fn(torch.randn(16, 7))
+        self.assertEqual(len(backend.graphs), 3)
+
+    @torch._dynamo.config.patch(automatic_dynamic_shapes=True)
+    def test_tensorspec_none_entry_inherits_automatic_dynamic(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:522]: keep it

**Review by laithsakka — APPROVED** [review]

File: test/dynamo/test_dynamic_spec.py (hunk 9 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py:522]: sg!

File: torch/_dynamo/dynamic_spec.py (hunk 5 above)
Comment (fxdawnn) [reply torch/_dynamo/dynamic_spec.py]: updated.

File: test/dynamo/test_dynamic_spec.py (hunk 8 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py]: deleted.
